File: 0146-lru-cache/0146-lru-cache.cpp

class LRUCache
{
public:
    int cap;
    struct Node
    {
        int key, value;
        Node *next, *prev;
        Node(int k, int v)
        {
            key = k;
            value = v;
            next = NULL;
            prev = NULL;
        }
    };
    map<int, Node *> mp;
    Node *head = new Node(-1, -1);
    Node *tail = new Node(-1, -1);
    LRUCache(int capacity)
    {
        cap = capacity;
        head->next = tail;
        tail->prev = head;
    }

    void deleteNode(Node *temp)
    {
        Node *back = temp->prev;
        Node *front = temp->next;
        back->next = front;
        front->prev = back;
    }

    void addNode(int key, int value)
    {
        Node *newNode = new Node(key, value);
        newNode->next = head->next;
        newNode->prev = head;
        head->next->prev = newNode;
        head->next = newNode;

        mp[key] = newNode;
    }

    int get(int key)
    {
        auto it = mp.find(key);
        if (it != mp.end())
        {
            Node *temp = it->second;
            deleteNode(temp);
            mp.erase(key);
            put(temp->key, temp->value);
            return temp->value;
        }
        else
            return -1;
    }

    void put(int key, int value)
    {
        auto it = mp.find(key);
        if (it != mp.end())
        {
            Node *temp = it->second;
            deleteNode(temp);
            mp.erase(temp->key);
            addNode(key, value);
        }
        else
        {
            if (mp.size() >= cap)
            {
                Node *leastUsed = tail->prev;
                deleteNode(leastUsed);
                mp.erase(leastUsed->key);
            }
            addNode(key, value);
        }
    }
};

File: 0146-lru-cache/0146-lru-cache.java
class DLL {
    int key;
    int val;
    DLL next;
    DLL prev;

    DLL(int key, int val) {
        this.key = key;
        this.val = val;
        this.next = null;
        this.prev = null;
    }
}

class LRUCache {
    HashMap<Integer, DLL> mp = new HashMap<>();
    DLL head;
    DLL tail;
    int capacity;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        head = new DLL(-1, -1); // dummy head
        tail = new DLL(-1, -1); // dummy tail
        head.next = tail;
        tail.prev = head;
    }

    void delete(DLL node) {
        DLL prevNode = node.prev;
        DLL nextNode = node.next;
        prevNode.next = nextNode;
        nextNode.prev = prevNode;
        node.prev = null;
        node.next = null;
    }

    void add(DLL node) {
        DLL nextTohead = head.next;
        node.next = nextTohead;
        node.prev = head;
        head.next = node;
        nextTohead.prev = node;
    }

    public int get(int key) {
        if (mp.containsKey(key)) {
            DLL d = mp.get(key);
            delete(d);
            add(d);
            return d.val;
        }
        return -1;
    }

    public void put(int key, int value) {
        if (mp.containsKey(key)) {
            // Update value and move to front
            DLL node = mp.get(key);
            node.val = value;
            delete(node);
            add(node);
        } else {
            if (mp.size() == capacity) {
                // Remove least recently used
                DLL leastUsedNode = tail.prev;
                delete(leastUsedNode);
                mp.remove(leastUsedNode.key);
            }
            DLL newNode = new DLL(key, value);
            add(newNode);
            mp.put(key, newNode);
        }
    }
}
